hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/dstable/DistributionSetTable.java

```diff
@@ -230,26 +231,55 @@ protected boolean isDropValid(final DragAndDropEvent dragEvent) {
         return super.isDropValid(dragEvent);
     }
 
-    private void assignSwmToDs(final Table source, final Set<Long> softwareModulesIdList, final long distId) {
+    private void handleSmToDsAssignment(final Set<Long> softwareModulesIdList, final long distId) {
         final Optional<DistributionSet> distributionSet = distributionSetManagement.get(distId);
-        distributionSet.ifPresent(set -> {
-            final DistributionSetIdName distributionSetIdName = new DistributionSetIdName(set);
-            selectDroppedEntities(distributionSetIdName.getId());
-            final HashMap<Long, HashSet<SoftwareModuleIdName>> map = createAssignmentMap(distributionSetIdName);
-            handleSoftwareModulesForAssignment(source, softwareModulesIdList, distId, map);
-            final HashSet<SoftwareModuleIdName> softwareModules = new HashSet<>();
-            map.keySet().forEach(typeId -> softwareModules.addAll(map.get(typeId)));
-            manageDistUIState.getAssignedList().put(distributionSetIdName, softwareModules);
-            openConfirmationWindowForAssignment(distributionSetIdName.getName(),
-                    softwareModules.toArray(new SoftwareModuleIdName[softwareModules.size()]));
-        });
         if (!distributionSet.isPresent()) {
             getNotification().displayWarning(getI18n().getMessage("distributionset.not.exists"));
+            return;
+        }
+        distributionSet.ifPresent(ds -> {
+            selectDroppedEntities(ds.getId());
+            final Set<SoftwareModule> softwareModules = getAssignableSoftwareModules(softwareModulesIdList, ds);
+            if (softwareModules.isEmpty()) {
+                return;
+            }
+            assignSoftwareModulesToDs(softwareModules, ds);
```

Comment:
You could reduce this to
            openConfirmationWindowForAssignment(ds.getName(),
                    softwareModules.stream().map(SoftwareModuleIdName::new)
                            .toArray(size -> new SoftwareModuleIdName[size]));

Reply:
fixed